src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts

```diff
@@ -0,0 +1,299 @@
+                if (!fs.lstatSync(remoteMachineTrailConfig.codeDir)
+                    .isDirectory()) {
+                    throw new Error(`codeDir ${remoteMachineTrailConfig.codeDir} is not a directory`);
+                }
+                try {
+                    // Validate to make sure codeDir doesn't have too many files
+                    await validateCodeDir(remoteMachineTrailConfig.codeDir);
+                } catch (error) {
+                    this.log.error(error);
+                    return Promise.reject(new Error(error));
+                }
+
+                this.trialConfig = remoteMachineTrailConfig;
+                break;
+            }
+            default:
+                this.log.debug(`Remote not support metadata key: '${key}', value: '${value}'`);
+        }
+    }
+
+    private scheduleMachine(): RemoteMachineMeta | undefined {
+        for (const [rmMeta, occupied] of this.remoteMachineMetaOccupiedMap) {
+            if (!occupied) {
+                this.remoteMachineMetaOccupiedMap.set(rmMeta, true);
+                return rmMeta;
+            }
+        }
+        return undefined;
+    }
+
+    private async setupConnections(machineList: string): Promise<void> {
+        this.log.debug(`Connecting to remote machines: ${machineList}`);
+        //TO DO: verify if value's format is wrong, and json parse failed, how to handle error
+        const rmMetaList: RemoteMachineMeta[] = <RemoteMachineMeta[]>JSON.parse(machineList);
+
+        for (const rmMeta of rmMetaList) {
+            this.sshConnectionPromises.push(await this.initRemoteMachineOnConnected(rmMeta));
+        }
+    }
+
+    private async initRemoteMachineOnConnected(rmMeta: RemoteMachineMeta): Promise<void> {
+        const executorManager: ExecutorManager = new ExecutorManager(rmMeta);
+        this.log.info(`connecting to ${rmMeta.username}@${rmMeta.ip}:${rmMeta.port}`);
+        const executor: ShellExecutor = await executorManager.getExecutor(this.initExecutorId);
+        this.log.debug(`reached ${executor.name}`);
+        this.machineExecutorManagerMap.set(rmMeta, executorManager);
+        this.log.debug(`initializing ${executor.name}`);
+
+        // Create root working directory after executor is ready
+        const nniRootDir: string = executor.joinPath(executor.getTempPath(), 'nni');
+        await executor.createFolder(executor.getRemoteExperimentRootDir(getExperimentId()));
+
+        // the directory to store temp scripts in remote machine
+        const remoteGpuScriptCollectorDir: string = executor.getRemoteScriptsPath(getExperimentId());
+
+        // clean up previous result.
+        await executor.createFolder(remoteGpuScriptCollectorDir, true);
+        await executor.allowPermission(true, nniRootDir);
+    }
+
+    public async refreshEnvironmentsStatus(environments: EnvironmentInformation[]): Promise<void> {
+        environments.forEach(async (environment) => {
```

Comment:
is it better to refresh all environment concurrently?

Reply:
updated